quickwit-storage/src/bundle_storage.rs

```diff
@@ -0,0 +1,276 @@
+            &self.bundle_file_name, self.metadata
+        )
+    }
+}
+
+fn unsupported_operation(path: &Path) -> StorageError {
+    let msg = "Unsupported operation. BundleStorage only supports async reads";
+    error!(path=?path, msg);
+    io::Error::new(io::ErrorKind::Other, format!("{}: {:?}", msg, path)).into()
+}
+
+/// BundleStorage bundles together multiple files into a single file
+/// with some metadata
+pub struct BundleStorageBuilder {
+    metadata: BundleStorageMetadata,
+    current_offset: usize,
+    /// The offset in the file where the hotcache begins. This is used to read
+    /// the hotcache and footer in a single read.
+    hotcache_offset: usize,
+    /// The footer offset, where the footer begins.
+    /// The footer includes the footer metadata as json encoded and the size of the footer in
+    /// bytes.
+    footer_offset: usize,
+    bundle_file: File,
+}
+
+impl BundleStorageBuilder {
+    /// Creates a new BundleStorageBuilder, to which files can be added.
+    pub fn new(path: &Path) -> io::Result<Self> {
+        let sink = File::create(path)?;
+
+        Ok(BundleStorageBuilder {
+            bundle_file: sink,
+            current_offset: 0,
+            hotcache_offset: 0,
+            footer_offset: 0,
+            metadata: Default::default(),
+        })
+    }
+
+    /// Appends a file to the bundle file.
+    ///
+    /// The hotcache needs to be the last file that is added, in order to be able to read
+    /// the hotcache and the metadata in one continous read.
+    pub fn add_file(&mut self, path: &Path) -> io::Result<()> {
+        let mut file = File::open(path)?;
+        let bytes_written = io::copy(&mut file, &mut self.bundle_file)? as usize;
+        let file_name = PathBuf::from(path.file_name().ok_or_else(|| {
+            io::Error::new(
+                ErrorKind::InvalidInput,
+                format!("could not extract file_name from path {:?}", path),
+            )
+        })?);
+        if file_name.to_string_lossy() == HOTCACHE_FILENAME {
+            self.hotcache_offset = self.current_offset;
+        }
+        self.metadata.files.insert(
+            file_name,
+            self.current_offset..self.current_offset + bytes_written,
+        );
+        self.current_offset += bytes_written;
+        Ok(())
+    }
+
+    /// Writes the metadata into the footer.
+    ///
+    pub fn finalize(mut self) -> io::Result<(usize, usize)> {
+        self.footer_offset = self.current_offset;
+        let metadata_json = serde_json::to_string(&self.metadata)?;
+        self.bundle_file.write_all(metadata_json.as_bytes())?;
+        let len = metadata_json.len() as u64;
+        BinarySerializable::serialize(&len, &mut self.bundle_file)?;
+        Ok((self.footer_offset, self.hotcache_offset))
+    }
+}
+#[cfg(test)]
+mod tests {
+    use std::env::temp_dir;
+
+    use crate::LocalFileStorage;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn bundlestorage_test() -> anyhow::Result<()> {
```

Comment:
Can we add a silly test for what happens if the bundlestorage contains no file?

Reply:
do you mean if we add no file to the bundle or if the underlying file does not exist?

Reply:
If we add no file to the bundle. NOt that I think there is a bug, but that is a good black box test to have.